packages/base-controller/src/BaseControllerV2.ts

```diff
@@ -250,29 +258,30 @@ function deriveStateFromMetadata<S extends Record<string, Json>>(
   metadata: StateMetadata<S>,
   metadataProperty: 'anonymous' | 'persist',
 ): Record<string, Json> {
-  return Object.keys(state).reduce((persistedState, key) => {
-    try {
-      const stateMetadata = metadata[key as keyof S];
-      if (!stateMetadata) {
-        throw new Error(`No metadata found for '${key}'`);
+  return Object.keys(state).reduce(
+    (persistedState: Record<string, Json>, key: keyof S) => {
+      try {
+        const stateMetadata = metadata[key];
+        if (!stateMetadata) {
+          throw new Error(`No metadata found for '${String(key)}'`);
+        }
+        const propertyMetadata = stateMetadata[metadataProperty];
+        const stateProperty = state[key];
+        if (typeof propertyMetadata === 'function') {
+          persistedState[String(key)] = propertyMetadata(stateProperty);
+        } else if (propertyMetadata) {
+          persistedState[String(key)] = stateProperty;
+        }
+        return persistedState;
+      } catch (error) {
+        // Throw error after timeout so that it is captured as a console error
+        // (and by Sentry) without interrupting state-related operations
+        setTimeout(() => {
+          throw error;
+        });
```

Comment:
I know this was originally here, but this seems very hacky 😅 Is there no other way to capture an error in Sentry, without throwing?

Reply:
We could accept a `captureError` method in the constructor, but IMO this solution is a lot nicer. No need for extra required constructor params for every single controller. Relying on global error handlers is framework-agnostic too.

We could hide this behind a utility function. Not sure we'll need it in very many places though.